src/main/java/io/cdap/plugin/servicenow/source/ServiceNowBaseSourceConfig.java

```diff
@@ -0,0 +1,318 @@
+    return password;
+  }
+
+  @Nullable
+  public String getStartDate() {
+    return startDate;
+  }
+
+  public String getEndDate() {
+    return endDate;
+  }
+
+  @Nullable
+  public String getTableNameField() {
+    return Strings.isNullOrEmpty(tableNameField) ? TABLE_NAME_FIELD_DEFAULT : tableNameField;
+  }
+
+  /**
+   * Validates {@link ServiceNowSourceConfig} instance.
+   */
+  public void validate(FailureCollector collector) {
+    // Validates the given referenceName to consists of characters allowed to represent a dataset.
+    IdUtils.validateReferenceName(referenceName, collector);
+
+    validateCredentials(collector);
+    validateValueType(collector);
+    validateDateRange(collector);
+    validateTableNameField(collector);
+  }
+
+  public void validateCredentials(FailureCollector collector) {
+    if (!shouldConnect()) {
+      return;
+    }
+
+    if (Util.isNullOrEmpty(clientId)) {
+      collector.addFailure("Client ID must be specified.", null)
+        .withConfigProperty(ServiceNowConstants.PROPERTY_CLIENT_ID);
+    }
+
+    if (Util.isNullOrEmpty(clientSecret)) {
+      collector.addFailure("Client Secret must be specified.", null)
+        .withConfigProperty(ServiceNowConstants.PROPERTY_CLIENT_SECRET);
+    }
+
+    if (Util.isNullOrEmpty(restApiEndpoint)) {
+      collector.addFailure("API Endpoint must be specified.", null)
+        .withConfigProperty(ServiceNowConstants.PROPERTY_API_ENDPOINT);
+    }
+
+    if (Util.isNullOrEmpty(user)) {
+      collector.addFailure("User name must be specified.", null)
+        .withConfigProperty(ServiceNowConstants.PROPERTY_USER);
+    }
+
+    if (Util.isNullOrEmpty(password)) {
+      collector.addFailure("Password must be specified.", null)
+        .withConfigProperty(ServiceNowConstants.PROPERTY_PASSWORD);
+    }
+
+    validateServiceNowConnection(collector);
+  }
+
+
+  @VisibleForTesting
+  void validateServiceNowConnection(FailureCollector collector) {
+    try {
+      ServiceNowTableAPIClientImpl restApi = new ServiceNowTableAPIClientImpl(this);
+      restApi.getAccessToken();
+    } catch (Exception e) {
+      collector.addFailure("Unable to connect to ServiceNow Instance.",
+          "Ensure properties like Client ID, Client Secret, API Endpoint, User Name, Password " +
+            "are correct.")
+        .withConfigProperty(ServiceNowConstants.PROPERTY_CLIENT_ID)
+        .withConfigProperty(ServiceNowConstants.PROPERTY_CLIENT_SECRET)
+        .withConfigProperty(ServiceNowConstants.PROPERTY_API_ENDPOINT)
+        .withConfigProperty(ServiceNowConstants.PROPERTY_USER)
+        .withConfigProperty(ServiceNowConstants.PROPERTY_PASSWORD)
+        .withStacktrace(e.getStackTrace());
+    }
+  }
+
+  /**
+   * Returns the value type chosen.
+   *
+   * @param collector The failure collector to collect the errors
+   * @return An instance of SourceValueType
+   */
+  public SourceValueType getValueType(FailureCollector collector) {
```

Comment:
If this method is not getting used outside the class, make it private

Reply:
No, it's being used by test classes.

Reply:
Changed to package-private and added the annotation: ` @VisibleForTesting` 69415df